src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -151,29 +152,31 @@ protected void enqueueOperation(Operation operation) {
 
   @Override
   protected void updateOperationWatchers(Operation operation) {
-    List<Function<Operation, Boolean>> operationWatchers =
-        watchers.get(operation.getName());
-    if (operationWatchers != null) {
-      if (operation.getDone()) {
-        watchers.remove(operation.getName());
-      }
-      long unfilteredWatcherCount = operationWatchers.size();
-      super.updateOperationWatchers(operation);
-      ImmutableList.Builder<Function<Operation, Boolean>> filteredWatchers = new ImmutableList.Builder<>();
-      long filteredWatcherCount = 0;
-      for (Function<Operation, Boolean> watcher : operationWatchers) {
-        if (watcher.apply(operation)) {
-          filteredWatchers.add(watcher);
-          filteredWatcherCount++;
+    synchronized(watchers) {
```

Comment:
Also, consider documenting on the watchers field that accesses need to be synchronized.